### PR TITLE
[`ruff`] Unnecessary cast to `int` (`RUF046`)

### LINKED ISSUE #11412: Add a rule to catch int(round(...)
`int(round(x))`and  `int(round(x, 0))` and `int(round(x, None))` can all simply be written `round(x)`.  However, unnecessary casts to `int` are [prevalent](https://sourcegraph.com/search?q=context%3Aglobal+int%28round%28&patternType=keyword&sm=0&filters=%5B%5B%22lang%22%2C%22Python%22%2C%22lang%3Apython%22%5D%5D).  Please consider adding a rule to catch these?

### PR DESCRIPTION
**Pull request by InSyncWithFoo**
## Summary

Resolves #11412.

## Test Plan

`cargo nextest run` and `cargo insta test`.

### CODE REVIEW

**Comment by sbrugman** [comment]
+1 unnecessary cast to int. This is more future proof: when type inference is available unnecessary int casts could be detected for a much wider class of functions.

Consider expanding this rule to other builtin functions such as “len”.

File: crates/ruff_linter/src/rules/ruff/rules/unnecessary_round_cast.rs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        return None;
+    };
+
+    if !semantic.match_builtin_expr(func, "round") {
+        return None;
+    }
+
+    if arguments.len() != 1 && arguments.len() != 2 {
+        return None;
+    }
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/ruff/rules/unnecessary_round_cast.rs:106]: Nit: The 0 case is handled below by `find_argument("number", 0)?` and searching an empty `arguments` list is cheap

```suggestion
    if arguments.len() >2 {
        return None;
    }
```

File: crates/ruff_linter/src/rules/ruff/rules/unnecessary_cast_to_int.rs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+use ruff_text_size::TextRange;
+
+/// ## What it does
+/// Checks for `int` conversions of values that are already integers.
+///
+/// ## Why is this bad?
+/// Such a conversion is unnecessary.
+///
+/// ## Known problems
+/// This rule is prone to false positives due to type inference limitations.
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/ruff/rules/unnecessary_cast_to_int.rs:15]: We should add an example for the cases where this is the case. I would expect that the rule is correct in almost all cases

Comment (MichaReiser) [reply crates/ruff_linter/src/rules/ruff/rules/unnecessary_cast_to_int.rs:15]: That's good to know. 

We should extend the documentation and list all possible false positives and why they're false positives.

Comment (MichaReiser) [reply crates/ruff_linter/src/rules/ruff/rules/unnecessary_cast_to_int.rs:15]: Thanks

Comment (MichaReiser) [reply crates/ruff_linter/src/rules/ruff/rules/unnecessary_cast_to_int.rs:15]: I'd probably rewrite this too, to make it sound less intimidating. 

```
When values incorrectly override the `__round__`, `__ceil__`, `__floor__`,
or `__trunc__` operators such that they don't return an integer,
this rule may produce false positives.
```

@AlexWaygood I'm leaning towards marking the fix as safe even in those cases. It's correct that this *could* alter behavior but the problem isn't the fix but the non-API compatible override of the said dunder methods.

**Review by MichaReiser — CHANGES_REQUESTED** [review]
Thanks. 

The special casing where `ndigitis=0` is incorrect. We should not flag that usage.

File: crates/ruff_linter/src/rules/ruff/rules/unnecessary_cast_to_int.rs (hunk 2 above)
Comment (AlexWaygood) [reply crates/ruff_linter/src/rules/ruff/rules/unnecessary_cast_to_int.rs:15]: Yes, if you've implemented the methods correctly, this should be safe. I don't think we need to worry too much about edge cases where the user has implemented the dunder method in an incorrect way.

**Comment by NeilGirdhar** [comment]
> when type inference is available unnecessary int casts could be detected for a much wider class of functions.

Just FYI, this may not be possible because Boolean values are (unfortunately) integers:
```
def f(x: int):
  assert isinstance(x, int)
  print(x, int(x))  # If you remove `int` here, you change things.
f(True)
```

You could detect unnecessary casts, but you have to check that it's not used in any way that would matter, which is a lot trickier.

**Comment by NeilGirdhar** [comment]
> @NeilGirdhar I made sure to take this into account: All functions listed there either delegate to their corresponding dunder methods or return a strict instance of `int`. If you can find one that doesn't, please let me know.

Yes, I wasn't talking about this excellent PR 😄 .  I was just responding to the comment that suggest that type inference can be used to discover unnecessary integer casts.

File: crates/ruff_linter/src/rules/ruff/rules/unnecessary_cast_to_int.rs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+use ruff_text_size::TextRange;
+
+/// ## What it does
+/// Checks for `int` conversions of values that are already integers.
+///
+/// ## Why is this bad?
+/// Such a conversion is unnecessary.
+///
+/// ## Known problems
+/// When values incorrectly override the `__round__`, `__ceil__`, `__floor__`,
```
Comment (dscorbett) [inline crates/ruff_linter/src/rules/ruff/rules/unnecessary_cast_to_int.rs:16]: Such overrides are not necessarily incorrect. [The documentation](https://docs.python.org/3/reference/datamodel.html#object.__round__) says “all these methods should return the value of the object truncated to an [`Integral`](https://docs.python.org/3/library/numbers.html#numbers.Integral) (typically an [`int`](https://docs.python.org/3/library/functions.html#int))”. They don’t need to return instances of `int` specifically, so calling `int` on the result might sometimes be necessary.

Comment (MichaReiser) [reply crates/ruff_linter/src/rules/ruff/rules/unnecessary_cast_to_int.rs:16]: Good point. I'm sorry, that I misguided you. In that case my suggestion is to:

* Change the fix back to unsafe for the cases where this applies
* Add a fix safety section and explain for which method calls the fix is unsafe.

Comment (AlexWaygood) [reply crates/ruff_linter/src/rules/ruff/rules/unnecessary_cast_to_int.rs:16]: Yeah, agreed. Sorry from me too!

File: crates/ruff_linter/src/rules/ruff/rules/unnecessary_cast_to_int.rs [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    }
+
+    let number = arguments.find_argument("number", 0)?;
+    let ndigits = arguments.find_argument("ndigits", 1);
+
+    let number_is_int = match number {
+        Expr::Name(name) => is_int(checker.semantic(), name),
+        Expr::NumberLiteral(ExprNumberLiteral { value, .. }) => matches!(value, Number::Int(..)),
+        _ => false,
+    };
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/ruff/rules/unnecessary_cast_to_int.rs:141]: Could we mark the fix as safe if `number_is_int` is `true`?

**Review by MichaReiser — APPROVED** [review]
This looks good to me. I pushed a few smaller nits, updated the documentation, and marked the fix for `round(2)` as safe. Let me know if that looks good to you and I'll merge

**Comment by DanielYang59** [comment]
I completely missed the party, but I'm amazed by how fast you respond, couldn't appreciate more

**Pull request merged**